### DISCUSSION TITLE
Understanding the FIFO delivery nature (also with Groups)

### DISCUSSION THREAD

**Discussion by thetutlage** [taskforcesh/bullmq#924] [Q&A] [answered]
The [docs](https://docs.bullmq.io/guide/jobs/fifo) mention that FIFO queue processes jobs in the order they are received.

Does this mean, the jobs are given to the worker in the FIFO order or is their processing is guaranteed to be in order? Here's an example.

- I have 2 workers. Worker 1 and Worker 2 and each workers processes only one job at a time
- Next I have 4 jobs. Job1, Job2, Job3 and Job4
- The worker 1 gets the first job and needs 20seconds to process it
- The worker 2 gets the second job, realises that this job cannot be processed and puts it back to the queue (maybe by raising an exception).
- Now worker 2 is free. Will it get the Job3 or will it get Job2 again? If it gets Job3, then my assumption of FIFO is correct. If it gets Job2, then the FIFO nature is same as SQS.

Also, if I use groups (from the pro version). And put first 3 jobs in Group1. Will they be processed as explained above. Job1, Job2 (failed), then Job3 processed before Job2.

**Comment by manast** [comment, marked as answer]
> * The worker 2 gets the second job, realises that this job cannot be processed and puts it back to the queue (maybe by raising an exception).

So the way a job gets back to wait is normally via a retry, so whether Job3 gets processed next depends on how fast the Job2 is placed back to the wait status, but most likely it will process Job3 first in this scenario. In any case, there is no logic in the queue where Job3 would wait for Job2 to exit the queue before processing Job3 so to speak.


> Also, if I use groups (from the pro version). And put first 3 jobs in Group1. Will they be processed as explained above. Job1, Job2 (failed), then Job3 processed before Job2.

Yes. But if you have more groups then of course the jobs will be interleaved.

**Reply by thetutlage** [reply]
Thanks for the quick response.

Yeah I think it's really hard to achieve the scenario I described using the available Redis data structures.
